handler.go

```diff
@@ -1,13 +1,118 @@
+	return InstrumentHandlerFuncWithOpts(
+		HttpHandlerOpts{
+			Subsystem:   "http",
+			HandlerName: handlerName,
+		},
+		handlerFunc,
+	)
+}
+
+func InstrumentHandlerWithOpts(opts HttpHandlerOpts, handler http.Handler) http.HandlerFunc {
+	return InstrumentHandlerFuncWithOpts(opts, handler.ServeHTTP)
+}
+
+// InstrumentHandlerFuncWithOpts instruments http handler with in-flight gauge,
+// request count, handle duration, request size, and response size
+func InstrumentHandlerFuncWithOpts(opts HttpHandlerOpts, handlerFunc func(http.ResponseWriter, *http.Request)) http.HandlerFunc {
+	inFlightGauge := prometheus.NewGauge(prometheus.GaugeOpts{
+		Namespace: opts.Namespace,
+		Subsystem: opts.Subsystem,
+		Name:      fmt.Sprintf("%s_in_flight_requests", opts.HandlerName),
+		Help:      fmt.Sprintf("In-flight HTTP requests of handler %s,", opts.HandlerName),
+	})
+
+	requestTotal := prometheus.NewCounterVec(
+		prometheus.CounterOpts{
+			Namespace: opts.Namespace,
+			Subsystem: opts.Subsystem,
+			Name:      fmt.Sprintf("%s_api_requests_total", opts.HandlerName),
+			Help:      fmt.Sprintf("Total number of HTTP requests made by handler %s", opts.HandlerName),
+		},
+		[]string{"code", "method"},
+	)
+
+	requestDurationHistogramOpts := prometheus.HistogramOpts{
+		Namespace:   opts.Namespace,
+		Subsystem:   opts.Subsystem,
+		Name:        fmt.Sprintf("%s_request_duration_seconds", opts.HandlerName),
+		Help:        fmt.Sprintf("A histogram of latencies for requests of handler %s.", opts.HandlerName),
+		Buckets:     opts.DurationBucket,
+		ConstLabels: prometheus.Labels{"handler": opts.HandlerName},
+	}
+	if len(requestDurationHistogramOpts.Buckets) == 0 {
+		requestDurationHistogramOpts.Buckets = defaultDurationBucket
+	}
+	requestDurationHistogram := prometheus.NewHistogramVec(requestDurationHistogramOpts, []string{"method"})
+
+	requestSizeHistogramOpts := prometheus.HistogramOpts{
+		Namespace: opts.Namespace,
+		Subsystem: opts.Subsystem,
+		Name:      fmt.Sprintf("%s_request_size_bytes", opts.HandlerName),
+		Help:      fmt.Sprintf("A histogram of response sizes for requests of handler %s.", opts.HandlerName),
+		Buckets:   opts.RequestSizeBucket,
+	}
+	if len(requestSizeHistogramOpts.Buckets) == 0 {
+		requestSizeHistogramOpts.Buckets = defaultRequestSizeBucket
+	}
+	requestSizeHistogram := prometheus.NewHistogramVec(requestSizeHistogramOpts, []string{})
+
+	responseSizeHistogramOpts := prometheus.HistogramOpts{
+		Namespace: opts.Namespace,
+		Subsystem: opts.Subsystem,
+		Name:      fmt.Sprintf("%s_response_size_bytes", opts.HandlerName),
+		Help:      fmt.Sprintf("A histogram of response sizes for requests of handler %s.", opts.HandlerName),
+		Buckets:   opts.ResponseSizeBucket,
+	}
+	if len(responseSizeHistogramOpts.Buckets) == 0 {
+		responseSizeHistogramOpts.Buckets = defaultResponseSizeBucket
+	}
+	responseSizeHistogram := prometheus.NewHistogramVec(responseSizeHistogramOpts, []string{})
```

Comment:
Do these need to define the buckets?

Reply:
The buckets can be defined in HttpHandlerOpts. If not provided, it uses the default bucket: defaultResponseSizeBucket  = prometheus.ExponentialBuckets(1024, 2, 22) //1K to 4G